File: lib/AbstractLayer/AbsLayerDidManager.cpp
//
//  AbsLayerDidManager.cpp
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

#include "AbsLayerDidManager.hpp"

#include <DidManager.h>
#include <WalletError.h>

int64_t AbsLayer_DidManager_CreateDid(int64_t obj, int index)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::DidManager>* didManager = reinterpret_cast<std::shared_ptr<elastos::DidManager>*>(obj);

  std::shared_ptr<elastos::Did>* did = new std::shared_ptr<elastos::Did>();
  int ret = (*didManager)->CreateDid(index,  did);
  if (ret != E_WALLET_C_OK) {
    return 0;
  }
  
    return reinterpret_cast<int64_t>(did);
}

int AbsLayer_DidManager_DidRecover(int64_t obj)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::DidManager>* didManager = reinterpret_cast<std::shared_ptr<elastos::DidManager>*>(obj);
  
  return (*didManager)->Recover();
}

void AbsLayer_DidManager_DestroyDidManager(int64_t obj)
{
  if (obj == 0) {
    return;
  }
  std::shared_ptr<elastos::DidManager>* didManager = reinterpret_cast<std::shared_ptr<elastos::DidManager>*>(obj);
  
  didManager->reset();
  delete didManager;
}

File: lib/AbstractLayer/AbsLayerDidManager.swift
//
//  AbsLayerDidManager.swift
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

extension AbstractLayer {
  final class DidManager {
    private init() {}

    public static func CreateDid(obj: Int64, index: Int) -> Int64 {
      let did = AbsLayer_DidManager_CreateDid(obj, Int32(index))
      return did
    }
    
    public static func DidRecover(obj: Int64) -> Int {
      return Int(AbsLayer_DidManager_DidRecover(obj))
    }
    
    public static func DestroyDidManager(obj: Int64) {
      AbsLayer_DidManager_DestroyDidManager(obj)
    }
    
  }
}

